src/serac/physics/solid_mechanics.hpp

```diff
@@ -1073,6 +1211,65 @@ class SolidMechanics<order, dim, Parameters<parameter_space...>, std::integer_se
         return (*residual_)(DifferentiateWRT<NUM_STATE_VARS + parameter_indices>{}, displacement_, zero_,
                             shape_displacement_, *parameters_[parameter_indices].state...);
       }...};
+
+  /**
+   * @brief Calculate a list of constrained dofs in the true displacement vector from a function that
+   * returns true if a physical coordinate is in the constrained set
+   *
+   * @param is_node_constrained A function that takes a point in physical space and returns true if the contained
+   * degrees of freedom should be constrained
+   * @param component A marker for whetehr
+   * @return An array of the constrained true dofs
+   */
+  mfem::Array<int> calculateConstrainedDofs(std::function<bool(const mfem::Vector&)> is_node_constrained,
+                                            std::optional<int>                       component = {})
+  {
+    // Get the nodal positions for the displacement vector in grid function form
+    mfem::ParGridFunction nodal_positions(&displacement_.space());
+    mesh_.GetNodes(nodal_positions);
+
+    const int        num_nodes = nodal_positions.Size() / dim;
+    mfem::Array<int> constrained_dofs;
+
+    for (int i = 0; i < num_nodes; i++) {
+      // Determine if this "local" node (L-vector node) is in the local true vector. I.e. ensure this node is not a
+      // shared node owned by another processor
+      if (nodal_positions.ParFESpace()->GetLocalTDofNumber(i) >= 0) {
+        mfem::Vector     node_coords(dim);
+        mfem::Array<int> node_dofs;
+        for (int d = 0; d < dim; d++) {
+          // Check if this component of the displacement vector is constrained
+          bool is_active_component = true;
+          if (component) {
+            if (*component != d) {
+              is_active_component = false;
+            }
+          }
+
+          if (is_active_component) {
+            // Get the local dof number for the prescribed component
+            int local_vector_dof = mfem::Ordering::Map<mfem::Ordering::byNODES>(
+                nodal_positions.FESpace()->GetNDofs(), nodal_positions.FESpace()->GetVDim(), i, d);
+
+            // Save the spatial position for this coordinate dof
+            node_coords(d) = nodal_positions(local_vector_dof);
```

Comment:
I think we want to initialize each component of the nodal position vector unconditionally